src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = { ...state };
+  const stateCopyArray = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const removeKey of action.keysToRemove) {
+          delete stateCopy[removeKey];
+        }
+        break;
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
+        break;
+      default:
+        break;
+    }
+
+    stateCopyArray.push(stateCopy);
+    stateCopy = { ...stateCopyArray[stateCopyArray.length - 1] };
```

Comment:
you can write it simpler
```suggestion
    stateCopyArray.push({ ...stateCopy });
```